File: python/82. Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        dummyhead = ListNode(0)
        dummyhead.next = head
        pre = dummyhead
        cur = pre.next
        next = cur.next
        while pre and cur and next:
            if cur.val == next.val:
                while next and cur.val == next.val:
                    next = next.next
                pre.next = next
                cur = next
                if next:
                    next = next.next
            else:
                pre = cur
                cur = next
                next = next.next
        return dummyhead.next

File: cpp/51-100/82. Remove Duplicates from Sorted List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	ListNode* deleteDuplicates(ListNode* head) {
		ListNode *dummyhead = new ListNode(-1);
		if (!head)
			return NULL;
		dummyhead->next = head;
		ListNode *p1 = dummyhead;
		ListNode *p2 = head->next;
		while(p2)
		{
			if(p1->next->val != p2->val){
				p1 = p1->next;
				p2 = p2->next;
			}
			else{
				while(p2 && p1->next->val == p2->val)
					p2 = p2->next;
				p1->next = p2;
				if(p2)
					p2 = p2->next;
			}
		}
		return dummyhead->next;
	}
};